src/resolve-celo.ts

```diff
@@ -0,0 +1,88 @@
+import Web3 from 'web3'
+import { Address } from '@celo/base/lib/address'
+import { AbiItem } from 'web3-utils'
+import { Contract } from 'web3-eth-contract'
+import { OdisUtils } from '@celo/identity'
+import { AuthSigner, ServiceContext } from '@celo/identity/lib/odis/query'
+
+import { NameResolver, NameResolutionResults, ResolutionKind } from './types'
+
+const DEFAULT_NETWORK_TIMEOUT = 5_000
+
+export class ResolveCelo implements NameResolver {
+  private federatedAttestationsContract: Contract
+  private authSigner: AuthSigner
+  private account: Address
+  private serviceContext: ServiceContext
+
+  static readonly FederatedAttestationsContractAbi: AbiItem[] =
+    require('./abis/FederatedAttestations.json').abi
+
+  // https://github.com/celo-org/identity/blob/ASv2/asv2/protocol.md#smart-contract-addresses
+  static readonly AlfajoresFederatedAttestationsProxyContractAddress: Address =
+    '0x70F9314aF173c246669cFb0EEe79F9Cfd9C34ee3'
+  static readonly MainnetFederatedAttestationsProxyContractAddress: Address =
+    '0x0aD5b1d0C25ecF6266Dd951403723B2687d6aff2'
+
+  constructor({
+    providerUrl,
+    federatedAttestationsProxyContractAddress,
+    authSigner,
+    account,
+    serviceContext,
+    networkTimeout,
+  }: {
+    providerUrl: string
+    federatedAttestationsProxyContractAddress: Address
+    authSigner: AuthSigner
+    account: Address
+    serviceContext: ServiceContext
+    networkTimeout?: number
+  }) {
+    networkTimeout = networkTimeout ?? DEFAULT_NETWORK_TIMEOUT
+    const web3 = new Web3(
+      new Web3.providers.HttpProvider(providerUrl, { timeout: networkTimeout }),
+    )
+    const federatedAttestationsContract = new web3.eth.Contract(
+      ResolveCelo.FederatedAttestationsContractAbi,
+      federatedAttestationsProxyContractAddress,
+    )
+
+    this.federatedAttestationsContract = federatedAttestationsContract
+    this.authSigner = authSigner
+    this.account = account
+    this.serviceContext = serviceContext
+  }
+
+  async resolve(id: string): Promise<NameResolutionResults> {
+    try {
+      // TODO(sbw): need a timeout
```

Comment:
you can just wrap the `getObfuscatedIdentifier` function in a timeout function `timeout(OdisUtils.Identifier.getObfuscatedIdentifier, [params], timeoutMs, ...) `

I would just note that the `getObfuscatedIdentifier` function wraps the ODIS query in a function that retries when the query returns an error, so the timeout would consider all the retries as one attempt

Reply:
...this might be dumb question: what's the `timeout` function?

Reply:
...I found a reference [1] in the celo-monorepo.  It looks like timouting this way during long-lived connections could cause cascading issues, like a bunch of open sockets building up.  We'd like to avoid that if possible.

[1]: https://github.com/celo-org/celo-monorepo/blob/master/packages/sdk/base/src/async.ts#L173

Reply:
from the slack thread surrounding this, it seems like we need to update either @celo/base or @celo/identity to take in an AbortController. We've created a ticket for this here and will work on implementing it ASAP - most likely in the new year https://github.com/celo-org/celo-monorepo/issues/10083

Reply:
updating that the above ticket is done ^